internal/provider/resource_id.go

```diff
@@ -29,152 +31,182 @@ the ` + "`create_before_destroy`" + ` lifecycle flag set to avoid conflicts with
+				PlanModifiers: []tfsdk.AttributePlanModifier{tfsdk.RequiresReplace()},
 			},
-
 			"b64_url": {
 				Description: "The generated id presented in base64, using the URL-friendly character set: " +
 					"case-sensitive letters, digits and the characters `_` and `-`.",
-				Type:     schema.TypeString,
+				Type:     types.StringType,
 				Computed: true,
 			},
-
 			"b64_std": {
 				Description: "The generated id presented in base64 without additional transformations.",
-				Type:        schema.TypeString,
+				Type:        types.StringType,
 				Computed:    true,
 			},
-
 			"hex": {
 				Description: "The generated id presented in padded hexadecimal digits. This result will " +
 					"always be twice as long as the requested byte length.",
-				Type:     schema.TypeString,
+				Type:     types.StringType,
 				Computed: true,
 			},
-
 			"dec": {
 				Description: "The generated id presented in non-padded decimal digits.",
-				Type:        schema.TypeString,
+				Type:        types.StringType,
 				Computed:    true,
 			},
-
 			"id": {
 				Description: "The generated id presented in base64 without additional transformations or prefix.",
-				Type:        schema.TypeString,
+				Type:        types.StringType,
 				Computed:    true,
 			},
 		},
-	}
+	}, nil
 }
 
-func CreateID(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	var diags diag.Diagnostics
-	byteLength := d.Get("byte_length").(int)
-	bytes := make([]byte, byteLength)
+func (r resourceIDType) NewResource(_ context.Context, p tfsdk.Provider) (tfsdk.Resource, diag.Diagnostics) {
+	return resourceID{
+		p: *(p.(*provider)),
+	}, nil
+}
 
-	n, err := rand.Reader.Read(bytes)
-	if n != byteLength {
-		return append(diags, diag.Errorf("generated insufficient random bytes: %s", err)...)
-	}
-	if err != nil {
-		return append(diags, diag.Errorf("error generating random bytes: %s", err)...)
-	}
+type resourceID struct {
+	p provider
+}
 
-	b64Str := base64.RawURLEncoding.EncodeToString(bytes)
-	d.SetId(b64Str)
+func (r resourceID) Create(ctx context.Context, req tfsdk.CreateResourceRequest, resp *tfsdk.CreateResourceResponse) {
+	var plan IDModel
 
-	repopEncsDiags := RepopulateEncodings(ctx, d, meta)
-	if repopEncsDiags != nil {
-		return append(diags, repopEncsDiags...)
+	diags := req.Plan.Get(ctx, &plan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
 	}
 
-	return diags
-}
-
-func RepopulateEncodings(_ context.Context, d *schema.ResourceData, _ interface{}) diag.Diagnostics {
-	var diags diag.Diagnostics
-	prefix := d.Get("prefix").(string)
-	base64Str := d.Id()
+	byteLength := plan.ByteLength.Value
+	bytes := make([]byte, byteLength)
 
-	bytes, err := base64.RawURLEncoding.DecodeString(base64Str)
+	n, err := rand.Reader.Read(bytes)
+	if int64(n) != byteLength {
+		resp.Diagnostics.AddError(
+			"generated insufficient random bytes: %s",
+			fmt.Sprintf("generated insufficient random bytes: %s", err),
+		)
```

Comment:
Nit: Diagnostic summaries should generally be a "title" of sorts, e.g. what's the higher level issue?, leaving the details to explain the actual problem, offer troubleshooting advice, and/or next steps. Similar comment applies to any other place that's copying contents between summary and details.

For something like this, it'd be great to say something like:

```suggestion
		resp.Diagnostics.AddError(
			"Randomness Generation Error",
			"While attempting to generate a random value for this resource, an insufficient number of random bytes were generated. Most commonly, this is a hardware or operating system issue where their random number generator does not provide a sufficient randomness source. Otherwise, it may represent an issue in the randomness handling of the provider.\n\n"+
				"Retry the Terraform operation. If the error still occurs or happens regularly, please contact the provider developer with hardware and operating system information.\n\n"+
				fmt.Sprintf("Original Error: %s", err)
		)
```

If desired, this diagnostic can then be wrapped in a quick function and reused for other resources.

Reference: https://www.terraform.io/plugin/framework/diagnostics

Reply:
I've used your wording and expanded the summary and detail messages for other diagnostics errors.